<a href="https://colab.research.google.com/github/xhdixhfl/kaggler/blob/main/breast_cancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 필요 모듈 & data import

In [20]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score

In [4]:
# datasets
bc = load_breast_cancer()

bc.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [6]:
# data explore
data = bc.data
print(data.shape)
print(data[0])

(569, 30)
[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]


In [9]:
label = bc.target
print(label.shape)
print(label)

print(bc.target_names) # 라벨이름
print(bc.DESCR) # 데이터 설명
print(bc.feature_names)  # feature에 대한 설명
print(bc.filename) # 파일이 저장된 경로

(569,)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0

# datasets split

In [12]:
import pandas as pd

df = pd.DataFrame(data, columns = bc.feature_names)
df['label'] = label# add target
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

In [13]:
X_train, X_test, y_train, y_test = train_test_split(data, label, test_size=0.2, random_state = 7)
print(len(X_test), len(X_train))
print(len(y_test), len(y_train))

114 455
114 455


# Decision Tree

In [14]:
from sklearn.tree import DecisionTreeClassifier

#model
decision_tree = DecisionTreeClassifier()
print(decision_tree._estimator_type)

# fit and predict
decision_tree.fit(X_train, y_train)
dt_pred = decision_tree.predict(X_test)

print(classification_report(y_test, dt_pred))

classifier
              precision    recall  f1-score   support

           0       0.95      0.88      0.91        40
           1       0.94      0.97      0.95        74

    accuracy                           0.94       114
   macro avg       0.94      0.92      0.93       114
weighted avg       0.94      0.94      0.94       114



# RandomForest

In [15]:
from sklearn.ensemble import RandomForestClassifier

#model
random_forest = RandomForestClassifier()
print(random_forest._estimator_type)

# fit and predict
random_forest.fit(X_train, y_train)
rf_pred = random_forest.predict(X_test)

print(classification_report(y_test, rf_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.90      0.95        40
           1       0.95      1.00      0.97        74

    accuracy                           0.96       114
   macro avg       0.97      0.95      0.96       114
weighted avg       0.97      0.96      0.96       114



In [19]:
from sklearn import svm

#svm model
svm_model = svm.SVC()
print(svm_model._estimator_type)

# fit and predict
svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)

print(classification_report(y_test, svm_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



In [18]:
from sklearn.linear_model import SGDClassifier

#sgd classifier 
sgd_model = SGDClassifier()
print(sgd_model._estimator_type)

# fit and predict
sgd_model.fit(X_train, y_train)
sgd_pred = sgd_model.predict(X_test)

print(classification_report(y_test, sgd_pred))

classifier
              precision    recall  f1-score   support

           0       0.66      0.95      0.78        40
           1       0.96      0.73      0.83        74

    accuracy                           0.81       114
   macro avg       0.81      0.84      0.80       114
weighted avg       0.86      0.81      0.81       114



In [17]:
from sklearn.linear_model import LogisticRegression

# model 
logistic_model = LogisticRegression()
print(logistic_model._estimator_type)

# fit and predict
logistic_model.fit(X_train, y_train)
lg_pred = logistic_model.predict(X_test)

print(classification_report(y_test, lg_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.82      0.90        40
           1       0.91      1.00      0.95        74

    accuracy                           0.94       114
   macro avg       0.96      0.91      0.93       114
weighted avg       0.94      0.94      0.94       114



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [21]:
#5개 모델 모두에 대해서 결과를 예측하고 성능을 확인
model_list = [decision_tree, random_forest, svm_model, sgd_model, logistic_model]
y_pred_list = []

for md in model_list:
    y_pred = md.predict(X_test)
    y_pred_list.append(y_pred)

acc_list = []
for pred in y_pred_list:
    acc_list.append(accuracy_score(y_test, pred))

print(acc_list)

[0.9385964912280702, 0.9649122807017544, 0.9035087719298246, 0.8070175438596491, 0.9385964912280702]


- 모델의 정확도는 랜덤포레스트의 결과가 0.96으로 가장 높은것을 알수 있다. 하지만, 유방암이라는 데이터 특성상 양성을 음성이라 판단하는 경우가 위험하기 때문에, 해당 지표를 알수 있는 recall을 활용하는게 적당하다.